// open-agents-nexus/dist/logger.js
const LEVELS = { debug: 0, info: 1, warn: 2, error: 3 };
let currentLevel = 'info';
export function setLogLevel(level) {
    currentLevel = level;
}
function shouldLog(level) {
    return LEVELS[level] >= LEVELS[currentLevel];
}
function format(level, scope, msg) {
    return `[${new Date().toISOString()}] [${level.toUpperCase()}] [${scope}] ${msg}`;
}
export function createLogger(scope) {
    return {
        debug: (msg, ...args) => {
            if (shouldLog('debug'))
                console.debug(format('debug', scope, msg), ...args);
        },
        info: (msg, ...args) => {
            if (shouldLog('info'))
                console.info(format('info', scope, msg), ...args);
        },
        warn: (msg, ...args) => {
            if (shouldLog('warn'))
                console.warn(format('warn', scope, msg), ...args);
        },
        error: (msg, ...args) => {
            if (shouldLog('error'))
                console.error(format('error', scope, msg), ...args);
        },
    };
}
