// @helia/utils/dist/src/utils/networked-storage.js
import { start, stop } from '@libp2p/interface';
import { InvalidConfigurationError } from "../errors.js";
import { SessionStorage } from "./session-storage.js";
import { Storage } from "./storage.js";
/**
 * Networked storage wraps a regular blockstore - when getting blocks if the
 * blocks are not present, the configured BlockBrokers will be used to fetch them.
 */
export class NetworkedStorage extends Storage {
    started;
    /**
     * Create a new BlockStorage
     */
    constructor(components, init = {}) {
        super(components, init);
        this.started = false;
    }
    isStarted() {
        return this.started;
    }
    async start() {
        await start(this.child, ...this.blockBrokers);
        this.started = true;
    }
    async stop() {
        await stop(this.child, ...this.blockBrokers);
        this.started = false;
    }
    unwrap() {
        return this.child;
    }
    createSession(root, options) {
        if (this.blockBrokers.length === 0) {
            throw new InvalidConfigurationError('No block brokers configured');
        }
        const blockBrokers = this.blockBrokers
            .map(broker => broker.createSession?.(options))
            .filter(broker => broker != null);
        if (blockBrokers.length === 0) {
            throw new InvalidConfigurationError(`No configured block brokers support sessions - tried ${this.blockBrokers.map(b => b.name).join(', ')}`);
        }
        return new SessionStorage({
            blockstore: this.child,
            blockBrokers,
            getHasher: this.getHasher,
            logger: this.logger
        }, {
            root
        });
    }
}
