// @helia/utils/dist/src/utils/session-storage.js
import { setMaxListeners } from '@libp2p/interface';
import { anySignal } from 'any-signal';
import { Storage } from "./storage.js";
/**
 * Storage subclass that can cancel any ongoing operation at any point.
 */
export class SessionStorage extends Storage {
    closeController;
    constructor(components, init) {
        super(components);
        // because brokers are allowed to continue searching for providers after the
        // session has been created, we need a way to tell them that the user has
        // finished using the session any in-flight requests should be cancelled
        this.closeController = new AbortController();
        setMaxListeners(Infinity, this.closeController.signal);
        this.log = components.logger.forComponent(`helia:session-storage:${init.root}`);
    }
    close() {
        this.closeController.abort();
    }
    async addPeer(peer, options) {
        await Promise.all(this.blockBrokers
            .map(broker => broker.addPeer(peer, options)));
    }
    /**
     * Put a block to the underlying datastore
     */
    async put(cid, block, options = {}) {
        const signal = anySignal([this.closeController.signal, options.signal]);
        setMaxListeners(Infinity, signal);
        try {
            return await super.put(cid, block, {
                ...options,
                signal
            });
        }
        finally {
            signal.clear();
        }
    }
    /**
     * Put a multiple blocks to the underlying datastore
     */
    async *putMany(blocks, options = {}) {
        const signal = anySignal([this.closeController.signal, options.signal]);
        setMaxListeners(Infinity, signal);
        try {
            yield* super.putMany(blocks, {
                ...options,
                signal
            });
        }
        finally {
            signal.clear();
        }
    }
    /**
     * Get a block by cid
     */
    async *get(cid, options = {}) {
        const signal = anySignal([this.closeController.signal, options.signal]);
        setMaxListeners(Infinity, signal);
        try {
            yield* super.get(cid, {
                ...options,
                signal
            });
        }
        finally {
            signal.clear();
        }
    }
    /**
     * Get multiple blocks back from an (async) iterable of cids
     */
    async *getMany(cids, options = {}) {
        const signal = anySignal([this.closeController.signal, options.signal]);
        setMaxListeners(Infinity, signal);
        try {
            yield* super.getMany(cids, {
                ...options,
                signal
            });
        }
        finally {
            signal.clear();
        }
    }
    /**
     * Delete a block from the blockstore
     */
    async delete(cid, options = {}) {
        const signal = anySignal([this.closeController.signal, options.signal]);
        setMaxListeners(Infinity, signal);
        try {
            await super.delete(cid, {
                ...options,
                signal
            });
        }
        finally {
            signal.clear();
        }
    }
    /**
     * Delete multiple blocks from the blockstore
     */
    async *deleteMany(cids, options = {}) {
        const signal = anySignal([this.closeController.signal, options.signal]);
        setMaxListeners(Infinity, signal);
        try {
            yield* super.deleteMany(cids, {
                ...options,
                signal
            });
        }
        finally {
            signal.clear();
        }
    }
    async has(cid, options = {}) {
        const signal = anySignal([this.closeController.signal, options.signal]);
        setMaxListeners(Infinity, signal);
        try {
            return await super.has(cid, {
                ...options,
                signal
            });
        }
        finally {
            signal.clear();
        }
    }
    async *getAll(options = {}) {
        const signal = anySignal([this.closeController.signal, options.signal]);
        setMaxListeners(Infinity, signal);
        try {
            yield* super.getAll({
                ...options,
                signal
            });
        }
        finally {
            signal.clear();
        }
    }
}
