// @helia/utils/dist/src/utils/storage.js
import { InvalidMultihashError, InvalidParametersError, setMaxListeners } from '@libp2p/interface';
import { anySignal } from 'any-signal';
import { IdentityBlockstore } from 'blockstore-core/identity';
import filter from 'it-filter';
import forEach from 'it-foreach';
import { CustomProgressEvent } from 'progress-events';
import { equals as uint8ArrayEquals } from 'uint8arrays/equals';
import { BlockNotFoundWhileOfflineError, InvalidConfigurationError, LoadBlockFailedError } from "../errors.js";
import { isPromise } from './is-promise.js';
const DEFAULT_MAX_IDENTITY_HASH_DIGEST_LENGTH = 128;
export class Storage {
    child;
    getHasher;
    log;
    logger;
    blockBrokers;
    /**
     * Create a new BlockStorage
     */
    constructor(components, init = {}) {
        this.log = components.logger.forComponent('helia:networked-storage');
        this.logger = components.logger;
        this.blockBrokers = components.blockBrokers;
        this.child = new IdentityBlockstore(components.blockstore, {
            maxDigestLength: init.maxIdentityHashDigestLength ?? DEFAULT_MAX_IDENTITY_HASH_DIGEST_LENGTH
        });
        this.getHasher = components.getHasher;
    }
    /**
     * Put a block to the underlying datastore
     */
    async put(cid, block, options = {}) {
        if (await this.child.has(cid, options)) {
            options.onProgress?.(new CustomProgressEvent('blocks:put:duplicate', cid));
            return cid;
        }
        options.onProgress?.(new CustomProgressEvent('blocks:put:providers:notify', cid));
        await Promise.all(this.blockBrokers.map(async (broker) => broker.announce?.(cid, options)));
        options.onProgress?.(new CustomProgressEvent('blocks:put:blockstore:put', cid));
        return this.child.put(cid, block, options);
    }
    /**
     * Put a multiple blocks to the underlying datastore
     */
    async *putMany(blocks, options = {}) {
        const missingBlocks = filter(blocks, async ({ cid }) => {
            const has = await this.child.has(cid, options);
            if (has) {
                options.onProgress?.(new CustomProgressEvent('blocks:put-many:duplicate', cid));
            }
            return !has;
        });
        const notifyEach = forEach(missingBlocks, async ({ cid }) => {
            options.onProgress?.(new CustomProgressEvent('blocks:put-many:providers:notify', cid));
            await Promise.all(this.blockBrokers.map(async (broker) => broker.announce?.(cid, options)));
        });
        options.onProgress?.(new CustomProgressEvent('blocks:put-many:blockstore:put-many'));
        yield* this.child.putMany(notifyEach, options);
    }
    /**
     * Get a block by cid
     */
    async *get(cid, options = {}) {
        const has = await this.child.has(cid, options);
        const offline = options.offline === true;
        if (!has) {
            if (offline) {
                throw new BlockNotFoundWhileOfflineError('The block was present in the blockstore and the node is running offline so cannot fetch it');
            }
            const hasher = await this.getHasher(cid.multihash.code);
            options?.signal?.throwIfAborted();
            // we do not have the block locally, get it from a block provider
            options.onProgress?.(new CustomProgressEvent('blocks:get:providers:get', cid));
            const block = await raceBlockRetrievers(cid, this.blockBrokers, hasher, {
                ...options,
                log: this.log
            });
            options.onProgress?.(new CustomProgressEvent('blocks:get:blockstore:put', cid));
            await this.child.put(cid, block, options);
            // notify other block providers of the new block
            options.onProgress?.(new CustomProgressEvent('blocks:get:providers:notify', cid));
            await Promise.all(this.blockBrokers.map(async (broker) => broker.announce?.(cid, options)));
            yield block;
            return;
        }
        options.onProgress?.(new CustomProgressEvent('blocks:get:blockstore:get', cid));
        yield* this.child.get(cid, options);
    }
    /**
     * Get multiple blocks back from an (async) iterable of cids
     */
    async *getMany(cids, options = {}) {
        options.onProgress?.(new CustomProgressEvent('blocks:get-many:blockstore:get-many'));
        yield* this.child.getMany(forEach(cids, async (cid) => {
            const has = await this.child.has(cid, options);
            const offline = options.offline === true;
            if (!has) {
                if (offline) {
                    throw new BlockNotFoundWhileOfflineError('The block was present in the blockstore and the node is running offline so cannot fetch it');
                }
                const hasher = await this.getHasher(cid.multihash.code);
                options?.signal?.throwIfAborted();
                // we do not have the block locally, get it from a block provider
                options.onProgress?.(new CustomProgressEvent('blocks:get-many:providers:get', cid));
                const block = await raceBlockRetrievers(cid, this.blockBrokers, hasher, {
                    ...options,
                    log: this.log
                });
                options.onProgress?.(new CustomProgressEvent('blocks:get-many:blockstore:put', cid));
                await this.child.put(cid, block, options);
                // notify other block providers of the new block
                options.onProgress?.(new CustomProgressEvent('blocks:get-many:providers:notify', cid));
                await Promise.all(this.blockBrokers.map(async (broker) => broker.announce?.(cid, options)));
            }
        }));
    }
    /**
     * Delete a block from the blockstore
     */
    async delete(cid, options = {}) {
        options.onProgress?.(new CustomProgressEvent('blocks:delete:blockstore:delete', cid));
        await this.child.delete(cid, options);
    }
    /**
     * Delete multiple blocks from the blockstore
     */
    async *deleteMany(cids, options = {}) {
        options.onProgress?.(new CustomProgressEvent('blocks:delete-many:blockstore:delete-many'));
        yield* this.child.deleteMany((async function* () {
            for await (const cid of cids) {
                yield cid;
            }
        }()), options);
    }
    async has(cid, options = {}) {
        return this.child.has(cid, options);
    }
    async *getAll(options = {}) {
        options.onProgress?.(new CustomProgressEvent('blocks:get-all:blockstore:get-many'));
        yield* this.child.getAll(options);
    }
}
/**
 * Race block providers cancelling any pending requests once the block has been
 * found.
 */
async function raceBlockRetrievers(cid, blockBrokers, hasher, options) {
    const validateFn = getCidBlockVerifierFunction(cid, hasher);
    const controller = new AbortController();
    const signal = anySignal([controller.signal, options.signal]);
    setMaxListeners(Infinity, controller.signal, signal);
    const retrievers = [];
    for (const broker of blockBrokers) {
        if (isRetrievingBlockBroker(broker)) {
            retrievers.push(broker);
        }
    }
    if (retrievers.length === 0) {
        throw new InvalidConfigurationError(`No block brokers capable of retrieving blocks are configured, the CID ${cid} cannot be fetched from the network`);
    }
    try {
        return await Promise.any(retrievers
            .map(async (retriever) => {
            try {
                let blocksWereValidated = false;
                const block = await retriever.retrieve(cid, {
                    ...options,
                    signal,
                    validateFn: async (block) => {
                        await validateFn(block);
                        options.signal?.throwIfAborted();
                        blocksWereValidated = true;
                    }
                });
                if (!blocksWereValidated) {
                    // the blockBroker either did not throw an error when attempting to validate the block
                    // or did not call the validateFn at all. We should validate the block ourselves
                    await validateFn(block);
                    options.signal?.throwIfAborted();
                }
                return block;
            }
            catch (err) {
                options.log.error('could not retrieve verified block for %c from %s - %e', cid, retriever.name, err);
                throw err;
            }
        }));
    }
    catch (err) {
        throw new LoadBlockFailedError(err.errors, `Failed to load block for ${cid}`);
    }
    finally {
        // we have the block from the fastest block retriever, abort any still
        // in-flight retrieve attempts
        controller.abort();
        signal.clear();
    }
}
function isRetrievingBlockBroker(broker) {
    return typeof broker.retrieve === 'function';
}
export const getCidBlockVerifierFunction = (cid, hasher) => {
    if (hasher == null) {
        throw new InvalidParametersError(`No hasher configured for multihash code 0x${cid.multihash.code.toString(16)}, please configure one. You can look up which hash this is at https://github.com/multiformats/multicodec/blob/master/table.csv`);
    }
    return async (block) => {
        // verify block
        let hash;
        const res = hasher.digest(block, {
            // support truncated hashes where they are truncated
            truncate: cid.multihash.digest.byteLength
        });
        if (isPromise(res)) {
            hash = await res;
        }
        else {
            hash = res;
        }
        if (!uint8ArrayEquals(hash.digest, cid.multihash.digest)) {
            // if a hash mismatch occurs for a TrustlessGatewayBlockBroker, we should try another gateway
            throw new InvalidMultihashError('Hash of downloaded block did not match multihash from passed CID');
        }
    };
};
