// @helia/bitswap/dist/src/peer-want-lists/index.js
import { trackedPeerMap } from '@libp2p/peer-collections';
import { CID } from 'multiformats/cid';
import { Ledger } from "./ledger.js";
export class PeerWantLists {
    blockstore;
    network;
    ledgerMap;
    maxSizeReplaceHasWithBlock;
    doNotResendBlockWindow;
    maxWantListSize;
    log;
    logger;
    constructor(components, init = {}) {
        this.blockstore = components.blockstore;
        this.network = components.network;
        this.maxSizeReplaceHasWithBlock = init.maxSizeReplaceHasWithBlock;
        this.doNotResendBlockWindow = init.doNotResendBlockWindow;
        this.maxWantListSize = init.maxWantListSize;
        this.log = components.logger.forComponent('helia:bitswap:peer-want-lists');
        this.logger = components.logger;
        this.ledgerMap = trackedPeerMap({
            name: 'helia_bitswap_ledger_map',
            metrics: components.metrics
        });
        this.network.addEventListener('bitswap:message', (evt) => {
            this.receiveMessage(evt.detail.peer, evt.detail.message)
                .catch(err => {
                this.log.error('error receiving bitswap message from %p - %e', evt.detail.peer, err);
            });
        });
        this.network.addEventListener('peer:disconnected', evt => {
            this.peerDisconnected(evt.detail);
        });
    }
    ledgerForPeer(peerId) {
        const ledger = this.ledgerMap.get(peerId);
        if (ledger == null) {
            return undefined;
        }
        return {
            peer: ledger.peerId,
            value: ledger.debtRatio(),
            sent: ledger.bytesSent,
            received: ledger.bytesReceived,
            exchanged: ledger.exchangeCount
        };
    }
    wantListForPeer(peerId) {
        const ledger = this.ledgerMap.get(peerId);
        if (ledger == null) {
            return undefined;
        }
        // remove any expired wants
        ledger.removeExpiredWants();
        return ledger.getWants();
    }
    peers() {
        return Array.from(this.ledgerMap.values()).map((l) => l.peerId);
    }
    /**
     * Handle incoming messages
     */
    async receiveMessage(peerId, message) {
        let ledger = this.ledgerMap.get(peerId);
        if (ledger == null) {
            ledger = new Ledger({
                peerId,
                blockstore: this.blockstore,
                network: this.network,
                logger: this.logger
            }, {
                maxSizeReplaceHasWithBlock: this.maxSizeReplaceHasWithBlock,
                doNotResendBlockWindow: this.doNotResendBlockWindow,
                maxWantListSize: this.maxWantListSize
            });
            this.ledgerMap.set(peerId, ledger);
        }
        // record the amount of block data received
        ledger.receivedBytes(message.blocks?.reduce((acc, curr) => acc + curr.data.byteLength, 0) ?? 0);
        // remove any expired wants
        ledger.removeExpiredWants();
        // add new wants
        ledger.addWants(message.wantlist);
        this.log('send blocks to peer');
        await ledger.sendBlocksToPeer();
    }
    async receivedBlock(cid, options) {
        const ledgers = [];
        for (const ledger of this.ledgerMap.values()) {
            if (ledger.hasWant(cid)) {
                ledgers.push(ledger);
            }
        }
        await Promise.all(ledgers.map(async (ledger) => ledger.sendBlocksToPeer(options)));
    }
    peerDisconnected(peerId) {
        this.ledgerMap.delete(peerId);
    }
}
