// @helia/bitswap/dist/src/session.js
import { AbstractSession } from '@helia/utils';
import { isPeerId } from '@libp2p/interface';
import { CustomProgressEvent } from 'progress-events';
class BitswapSession extends AbstractSession {
    name = 'bitswap-session';
    wantList;
    network;
    libp2p;
    constructor(components, init) {
        super(components, {
            ...init,
            name: 'helia:bitswap:session'
        });
        this.wantList = components.wantList;
        this.network = components.network;
        this.libp2p = components.libp2p;
    }
    async queryProvider(cid, provider, options) {
        this.log('sending WANT-BLOCK for %c to %p', cid, provider);
        const result = await this.wantList.wantSessionBlock(cid, provider.peerId, options);
        this.log('%p %s %c', provider, result.has ? 'has' : 'does not have', cid);
        if (result.has && result.block != null) {
            return result.block;
        }
        throw new Error('Provider did not have block');
    }
    async *findNewProviders(cid, options = {}) {
        for await (const provider of this.network.findProviders(cid, options)) {
            yield {
                peerId: provider.id,
                routing: provider.routing,
                toString: () => `Bitswap(${provider.id})`
            };
        }
    }
    toFilterKey(provider) {
        return provider.peerId.toMultihash().bytes;
    }
    equals(providerA, providerB) {
        return providerA.peerId.equals(providerB.peerId);
    }
    async convertToProvider(provider, routing, options) {
        if (isPeerId(provider)) {
            return {
                peerId: provider,
                routing,
                toString: () => `Bitswap(${provider})`
            };
        }
        if (await this.libp2p.isDialable(provider) === false) {
            return;
        }
        try {
            const connection = await this.libp2p.dial(provider, options);
            return {
                peerId: connection.remotePeer,
                routing,
                toString: () => `Bitswap(${connection.remotePeer})`
            };
        }
        catch { }
    }
    emitFoundProviderProgressEvent(cid, provider, options) {
        options?.onProgress?.(new CustomProgressEvent('bitswap:found-provider', {
            type: 'bitswap',
            cid,
            provider: {
                id: provider.peerId,
                multiaddrs: [],
                routing: provider.routing
            },
            routing: provider.routing
        }));
    }
}
export function createBitswapSession(components, init) {
    return new BitswapSession(components, init);
}
