// blockstore-fs/dist/src/index.js
/**
 * @packageDocumentation
 *
 * A Blockstore implementation that stores blocks in the local filesystem.
 *
 * @example
 *
 * ```js
 * import { FsBlockstore } from 'blockstore-fs'
 *
 * const store = new FsBlockstore('path/to/store')
 * ```
 */
import fs from 'node:fs/promises';
import path from 'node:path';
import { OpenFailedError, PutFailedError, NotFoundError, DeleteFailedError } from 'interface-store';
import glob from 'it-glob';
import map from 'it-map';
import parallelBatch from 'it-parallel-batch';
import { raceSignal } from 'race-signal';
import { Writer } from 'steno';
import { NextToLast } from './sharding.js';
/**
 * Write a file atomically
 */
async function writeFile(file, contents, options) {
    try {
        options?.signal?.throwIfAborted();
        await raceSignal(fs.mkdir(path.dirname(file), {
            recursive: true
        }), options?.signal);
        const writer = new Writer(file);
        options?.signal?.throwIfAborted();
        await writer.write(contents);
    }
    catch (err) {
        if (err.syscall === 'rename' && ['ENOENT', 'EPERM'].includes(err.code)) {
            // steno writes a file to a temp location before renaming it.
            // If the final file already exists this error is thrown.
            // Make sure we can read & write to this file
            options?.signal?.throwIfAborted();
            await raceSignal(fs.access(file, fs.constants.F_OK | fs.constants.W_OK), options?.signal);
            // The file was created by another context - this means there were
            // attempts to write the same block by two different function calls
            return;
        }
        throw err;
    }
}
/**
 * A blockstore backed by the file system
 */
export class FsBlockstore {
    path;
    createIfMissing;
    errorIfExists;
    putManyConcurrency;
    getManyConcurrency;
    deleteManyConcurrency;
    shardingStrategy;
    constructor(location, init = {}) {
        this.path = path.resolve(location);
        this.createIfMissing = init.createIfMissing ?? true;
        this.errorIfExists = init.errorIfExists ?? false;
        this.deleteManyConcurrency = init.deleteManyConcurrency ?? 50;
        this.getManyConcurrency = init.getManyConcurrency ?? 50;
        this.putManyConcurrency = init.putManyConcurrency ?? 50;
        this.shardingStrategy = init.shardingStrategy ?? new NextToLast();
    }
    async open() {
        try {
            await fs.access(this.path, fs.constants.F_OK | fs.constants.W_OK);
            if (this.errorIfExists) {
                throw new OpenFailedError(`Blockstore directory: ${this.path} already exists`);
            }
        }
        catch (err) {
            if (err.code === 'ENOENT') {
                if (this.createIfMissing) {
                    await fs.mkdir(this.path, { recursive: true });
                    return;
                }
                else {
                    throw new OpenFailedError(`Blockstore directory: ${this.path} does not exist`);
                }
            }
            throw err;
        }
    }
    async close() {
        await Promise.resolve();
    }
    async put(key, val, options) {
        const { dir, file } = this.shardingStrategy.encode(key);
        try {
            await writeFile(path.join(this.path, dir, file), val, options);
            return key;
        }
        catch (err) {
            throw new PutFailedError(String(err));
        }
    }
    async *putMany(source, options) {
        yield* parallelBatch(map(source, ({ cid, bytes }) => {
            return async () => {
                await this.put(cid, bytes, options);
                return cid;
            };
        }), this.putManyConcurrency);
    }
    async *get(key, options) {
        const { dir, file } = this.shardingStrategy.encode(key);
        let handle;
        try {
            options?.signal?.throwIfAborted();
            handle = await raceSignal(fs.open(path.join(this.path, dir, file)), options?.signal);
            yield* handle.createReadStream();
        }
        catch (err) {
            if (handle != null) {
                await raceSignal(handle.close(), options?.signal);
            }
            throw new NotFoundError(String(err));
        }
    }
    async *getMany(source, options) {
        yield* parallelBatch(map(source, key => {
            return async () => {
                return {
                    cid: key,
                    bytes: this.get(key, options)
                };
            };
        }), this.getManyConcurrency);
    }
    async delete(key, options) {
        const { dir, file } = this.shardingStrategy.encode(key);
        try {
            options?.signal?.throwIfAborted();
            await raceSignal(fs.unlink(path.join(this.path, dir, file)), options?.signal);
        }
        catch (err) {
            if (err.code === 'ENOENT') {
                return;
            }
            throw new DeleteFailedError(String(err));
        }
    }
    async *deleteMany(source, options) {
        yield* parallelBatch(map(source, key => {
            return async () => {
                await this.delete(key, options);
                return key;
            };
        }), this.deleteManyConcurrency);
    }
    async has(key, options) {
        const { dir, file } = this.shardingStrategy.encode(key);
        try {
            options?.signal?.throwIfAborted();
            await raceSignal(fs.access(path.join(this.path, dir, file)), options?.signal);
        }
        catch (err) {
            return false;
        }
        return true;
    }
    async *getAll(options) {
        const pattern = `**/*${this.shardingStrategy.extension}`
            .split(path.sep)
            .join('/');
        const files = glob(this.path, pattern, {
            absolute: true
        });
        for await (const file of files) {
            try {
                options?.signal?.throwIfAborted();
                const pair = {
                    cid: this.shardingStrategy.decode(file),
                    bytes: (async function* () {
                        let handle;
                        try {
                            handle = await raceSignal(fs.open(file), options?.signal);
                            yield* handle.createReadStream();
                        }
                        finally {
                            if (handle != null) {
                                await raceSignal(handle.close(), options?.signal);
                            }
                        }
                    })()
                };
                yield pair;
                options?.signal?.throwIfAborted();
            }
            catch (err) {
                // if keys are removed from the blockstore while the query is
                // running, we may encounter missing files.
                if (err.code !== 'ENOENT') {
                    throw err;
                }
            }
        }
    }
}
