// @libp2p/http/dist/src/http.browser.js
import { fetch } from '@libp2p/http-fetch';
import { getHeaders, getHost, stripHTTPPath, toResource } from '@libp2p/http-utils';
import { WebSocket } from '@libp2p/http-websocket';
import { UnsupportedOperationError, serviceCapabilities, start, stop } from '@libp2p/interface';
import { HTTP_PROTOCOL } from './constants.js';
import { Cookies } from './middleware/cookies.js';
import { Origin } from './middleware/origin.js';
import { HTTPRegistrar } from './registrar.js';
import { prepareAndConnect, prepareAndSendRequest, processResponse } from './utils.js';
import { WELL_KNOWN_PROTOCOLS_PATH } from './index.js';
export class HTTP {
    log;
    components;
    httpRegistrar;
    origin;
    cookies;
    constructor(components, init = {}) {
        this.components = components;
        this.log = components.logger.forComponent('libp2p:http');
        this.httpRegistrar = new HTTPRegistrar(components, init);
        this.origin = new Origin();
        this.cookies = new Cookies(components, init);
    }
    [Symbol.toStringTag] = '@libp2p/http';
    [serviceCapabilities] = [
        '@libp2p/http'
    ];
    async start() {
        await start(this.httpRegistrar);
    }
    async stop() {
        await stop(this.httpRegistrar);
    }
    agent(...args) {
        throw new UnsupportedOperationError('This method is not supported in browsers');
    }
    dispatcher(...args) {
        throw new UnsupportedOperationError('This method is not supported in browsers');
    }
    async connect(resource, init = {}) {
        const url = toResource(resource);
        const headers = getHeaders(init);
        const opts = {
            ...init,
            headers,
            method: 'GET',
            middleware: init.middleware?.map(fn => fn(this.components)) ?? []
        };
        headers.set('connection', 'upgrade');
        headers.set('upgrade', 'websocket');
        return prepareAndConnect(url, opts, async () => {
            if (url instanceof URL) {
                const socket = new globalThis.WebSocket(url, init.protocols);
                socket.binaryType = 'arraybuffer';
                return socket;
            }
            // strip http-path tuple but record the value if set
            const { addresses, httpPath } = stripHTTPPath(url);
            return new WebSocket(addresses, new URL(`http://${getHost(url, opts.headers)}${decodeURIComponent(httpPath)}`), this.components.connectionManager, opts);
        });
    }
    async fetch(resource, init = {}) {
        const url = toResource(resource);
        const opts = {
            ...init,
            headers: getHeaders(init),
            method: 'GET',
            middleware: [
                this.origin,
                this.cookies,
                ...init.middleware?.map(fn => fn(this.components)) ?? []
            ]
        };
        const response = await prepareAndSendRequest(url, opts, async () => {
            return this.sendRequest(url, init);
        });
        return processResponse(url, opts, response);
    }
    async connectProtocol(resource, protocol, init) {
        const path = await this.getProtocolPath(resource, protocol, init);
        const url = toResource(resource, path);
        return this.connect(url, init);
    }
    async fetchProtocol(resource, protocol, init = {}) {
        const path = await this.getProtocolPath(resource, protocol, init);
        const url = toResource(resource, path);
        return this.fetch(url, init);
    }
    async getSupportedProtocols(resource, options = {}) {
        const url = toResource(resource, WELL_KNOWN_PROTOCOLS_PATH);
        const resp = await this.fetch(url, {
            method: 'GET',
            headers: {
                Accept: 'application/json'
            },
            signal: options.signal
        });
        if (resp.status !== 200) {
            throw new Error(`Unexpected status code: ${resp.status}`);
        }
        return resp.json();
    }
    async getProtocolPath(peer, protocol, options = {}) {
        const peerMeta = await this.getSupportedProtocols(peer, options);
        if (peerMeta[protocol] == null) {
            throw new Error(`Peer does not serve protocol: ${protocol}`);
        }
        return peerMeta[protocol].path;
    }
    canHandle(req) {
        return this.httpRegistrar.canHandle(req);
    }
    async onRequest(req) {
        return this.httpRegistrar.onRequest(req);
    }
    onWebSocket(ws) {
        this.httpRegistrar.onWebSocket(ws);
    }
    handle(protocol, handler) {
        this.httpRegistrar.handle(protocol, handler);
    }
    unhandle(protocol) {
        this.httpRegistrar.unhandle(protocol);
    }
    getProtocolMap() {
        return this.httpRegistrar.getProtocolMap();
    }
    async sendRequest(resource, init) {
        if (resource instanceof URL) {
            this.log('making request to %s with global fetch');
            return globalThis.fetch(resource, init);
        }
        this.log('making request to %s with libp2p fetch', resource);
        const host = getHost(resource, getHeaders(init));
        // strip http-path tuple but record the value if set
        const { addresses, httpPath } = stripHTTPPath(resource);
        const connection = await this.components.connectionManager.openConnection(addresses, {
            signal: init.signal ?? undefined
        });
        const stream = await connection.newStream(HTTP_PROTOCOL, {
            signal: init.signal ?? undefined
        });
        return fetch(stream, new URL(`http://${host}${decodeURIComponent(httpPath)}`), init);
    }
}
