// @libp2p/http/dist/src/index.js
/**
 * @packageDocumentation
 *
 * This module allows you to use HTTP requests as a transport for libp2p
 * protocols (libp2p over HTTP), and also libp2p streams as a transport for HTTP
 * requests (HTTP over libp2p).
 *
 * It integrates with existing Node.js friendly HTTP frameworks such as
 * [express](https://expressjs.com/) and [Fastify](https://fastify.dev) as well
 * as [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request)/
 * [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response)-based
 * frameworks like [Hono](https://hono.dev/).
 *
 * It even allows creating Node.js-style [http.Server](https://nodejs.org/api/http.html#class-httpserver)s
 * and [WebSocketServer](https://github.com/websockets/ws/blob/HEAD/doc/ws.md#class-websocketserver)s
 * in browsers to truly realize the power of the distributed web.
 *
 * In addition to URL-based addressing, it can use a libp2p PeerId and/or
 * multiaddr(s) and lets libp2p take care of the routing, thus taking advantage
 * of features like multi-routes, NAT traversal and stream multiplexing over a
 * single connection.
 *
 * Check out the [examples](https://github.com/libp2p/js-libp2p-http/tree/main/examples)
 * for how you can use this in your applications!
 */
import { HTTP as HTTPClass } from './http.js';
export { WELL_KNOWN_PROTOCOLS_PATH } from './routes/well-known.js';
export { HTTP_PROTOCOL } from './constants.js';
/**
 * Create an HTTP service that provides a `fetch` implementation and a way to
 * register custom HTTP handlers.
 */
export function http(init = {}) {
    return (components) => new HTTPClass(components, init);
}
export { authenticatedRoute, authenticatedWebSocketRoute } from './routes/peer-id-auth.js';
export { webSocketRoute } from './routes/websocket.js';
