// open-agents-nexus/dist/x402/balance.js
/**
 * x402/balance.ts — COHERE Plane 3: Two-Balance Model
 *
 * Implements the dual-balance accounting system per unified COHERE paper (p.46).
 * Providers earn both cash-settled credits AND network access credits.
 * This lets providers either cash out or spend earned value back into the commons.
 *
 * Balance A: Cash-settled provider credits (USDC, withdrawable)
 * Balance B: Network access credits (internal, for consuming inference)
 *
 * Provenance: Project COHERE Unified Complete §Two-balance model
 */
import { createLogger } from '../logger.js';
const log = createLogger('balance');
// ---------------------------------------------------------------------------
// DualBalanceManager
// ---------------------------------------------------------------------------
export class DualBalanceManager {
    balance;
    ledger = [];
    maxLedgerSize;
    constructor(maxLedgerSize = 10000) {
        this.maxLedgerSize = maxLedgerSize;
        this.balance = {
            cashCredits: 0,
            networkCredits: 0,
            totalEarned: 0,
            totalSpentNetwork: 0,
            totalWithdrawn: 0,
            updatedAt: new Date().toISOString(),
        };
    }
    /** Earn from providing inference — splits into cash + network credits */
    earnFromInference(amount, counterparty, jobId) {
        // Default split: 80% cash, 20% network credits
        const cashPortion = amount * 0.80;
        const networkPortion = amount * 0.20;
        this.balance.cashCredits += cashPortion;
        this.balance.networkCredits += networkPortion;
        this.balance.totalEarned += amount;
        this.balance.updatedAt = new Date().toISOString();
        this.addTransaction({
            id: `txn-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 6)}`,
            type: 'earn_inference',
            balanceType: 'both',
            amount,
            counterparty,
            reference: jobId,
            description: `Earned $${amount.toFixed(6)} from inference (cash: $${cashPortion.toFixed(6)}, network: $${networkPortion.toFixed(6)})`,
            timestamp: new Date().toISOString(),
        });
        log.info(`Balance: earned $${amount.toFixed(6)} from ${counterparty} (job: ${jobId})`);
    }
    /** Spend network credits on remote inference */
    spendOnInference(amount, counterparty, jobId) {
        if (this.balance.networkCredits < amount) {
            log.warn(`Balance: insufficient network credits ($${this.balance.networkCredits.toFixed(6)} < $${amount.toFixed(6)})`);
            return false;
        }
        this.balance.networkCredits -= amount;
        this.balance.totalSpentNetwork += amount;
        this.balance.updatedAt = new Date().toISOString();
        this.addTransaction({
            id: `txn-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 6)}`,
            type: 'spend_inference',
            balanceType: 'network',
            amount: -amount,
            counterparty,
            reference: jobId,
            description: `Spent $${amount.toFixed(6)} on remote inference`,
            timestamp: new Date().toISOString(),
        });
        return true;
    }
    /** Withdraw cash credits to on-chain USDC */
    withdrawCash(amount) {
        if (this.balance.cashCredits < amount) {
            return false;
        }
        this.balance.cashCredits -= amount;
        this.balance.totalWithdrawn += amount;
        this.balance.updatedAt = new Date().toISOString();
        this.addTransaction({
            id: `txn-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 6)}`,
            type: 'withdraw_cash',
            balanceType: 'cash',
            amount: -amount,
            description: `Withdrew $${amount.toFixed(6)} to on-chain USDC`,
            timestamp: new Date().toISOString(),
        });
        log.info(`Balance: withdrew $${amount.toFixed(6)} to on-chain USDC`);
        return true;
    }
    /** Add contribution credits (for users sharing resources) */
    addContributionCredit(amount, type, peerId) {
        this.balance.networkCredits += amount;
        this.balance.updatedAt = new Date().toISOString();
        this.addTransaction({
            id: `txn-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 6)}`,
            type: 'contribution_credit',
            balanceType: 'network',
            amount,
            counterparty: peerId,
            description: `Contribution credit: ${type} ($${amount.toFixed(6)})`,
            timestamp: new Date().toISOString(),
        });
    }
    /** Get current balance */
    getBalance() {
        return { ...this.balance };
    }
    /** Get recent transactions */
    getTransactions(limit = 50) {
        return this.ledger.slice(-limit);
    }
    /** Check if user can afford a network spend */
    canAfford(amount) {
        return this.balance.networkCredits >= amount;
    }
    // ── Internal ──
    addTransaction(txn) {
        this.ledger.push(txn);
        if (this.ledger.length > this.maxLedgerSize) {
            this.ledger = this.ledger.slice(-Math.floor(this.maxLedgerSize * 0.8));
        }
    }
}
