// open-agents-nexus/dist/x402/verifier.js
/**
 * Self-Verified Payment Verification via Alchemy RPC
 *
 * Verifies EIP-3009 TransferWithAuthorization proofs by:
 * 1. Checking the EIP-712 signature matches the claimed payer
 * 2. Reading payer's USDC balance on Base via Alchemy
 * 3. Checking the authorization nonce hasn't been used (replay prevention)
 * 4. Validating timestamp bounds (validAfter / validBefore)
 *
 * No Coinbase facilitator — fully self-sovereign verification.
 */
import { createPublicClient, http } from 'viem';
import { base } from 'viem/chains';
import { verifyPaymentSignature, USDC_BASE_ADDRESS, } from './eip712.js';
import { createLogger } from '../logger.js';
const log = createLogger('x402:verifier');
// ---------------------------------------------------------------------------
// Minimal USDC ABI for read operations
// ---------------------------------------------------------------------------
const USDC_READ_ABI = [
    {
        name: 'balanceOf',
        type: 'function',
        stateMutability: 'view',
        inputs: [{ name: 'account', type: 'address' }],
        outputs: [{ name: '', type: 'uint256' }],
    },
    {
        name: 'authorizationState',
        type: 'function',
        stateMutability: 'view',
        inputs: [
            { name: 'authorizer', type: 'address' },
            { name: 'nonce', type: 'bytes32' },
        ],
        outputs: [{ name: '', type: 'bool' }],
    },
];
export class PaymentVerifier {
    client;
    constructor(client) {
        this.client = client;
    }
    /**
     * Create a verifier connected to Base mainnet via Alchemy.
     */
    static create(alchemyApiKey) {
        const client = createPublicClient({
            chain: base,
            transport: http(`https://base-mainnet.g.alchemy.com/v2/${alchemyApiKey}`),
        });
        return new PaymentVerifier(client);
    }
    /**
     * Read a payer's USDC balance on Base.
     */
    async checkBalance(address) {
        const balance = await this.client.readContract({
            address: USDC_BASE_ADDRESS,
            abi: USDC_READ_ABI,
            functionName: 'balanceOf',
            args: [address],
        });
        return balance;
    }
    /**
     * Check whether an EIP-3009 nonce has already been used.
     * Returns true if the nonce was consumed (i.e., replay detected).
     */
    async checkNonceUsed(authorizer, nonce) {
        return this.client.readContract({
            address: USDC_BASE_ADDRESS,
            abi: USDC_READ_ABI,
            functionName: 'authorizationState',
            args: [authorizer, nonce],
        });
    }
    /**
     * Full payment verification pipeline:
     * signature → timestamps → nonce → balance
     */
    async verify(from, message, signature) {
        // 1. Verify EIP-712 signature
        const sigValid = await verifyPaymentSignature(from, message, signature);
        if (!sigValid) {
            log.warn(`Signature verification failed for ${from}`);
            return { valid: false, reason: 'Invalid EIP-712 signature' };
        }
        // 2. Check timestamp bounds
        const now = BigInt(Math.floor(Date.now() / 1000));
        if (now < message.validAfter) {
            return { valid: false, reason: 'Authorization not yet valid', signerAddress: from };
        }
        if (now > message.validBefore) {
            return { valid: false, reason: 'Authorization expired', signerAddress: from };
        }
        // 3. Check nonce hasn't been replayed
        const nonceUsed = await this.checkNonceUsed(from, message.nonce);
        if (nonceUsed) {
            log.warn(`Nonce replay detected for ${from}`);
            return { valid: false, reason: 'Nonce already used (replay)', signerAddress: from };
        }
        // 4. Check payer has sufficient USDC balance
        const balance = await this.checkBalance(from);
        if (balance < message.value) {
            log.warn(`Insufficient balance: ${balance} < ${message.value} for ${from}`);
            return {
                valid: false,
                reason: 'Insufficient USDC balance',
                signerAddress: from,
                balance,
            };
        }
        log.info(`Payment verified: ${message.value} USDC from ${from}`);
        return { valid: true, signerAddress: from, balance };
    }
}
