// ipfs-unixfs-exporter/dist/src/exporters/unixfs-v1/index.js
import { decode } from '@ipld/dag-pb';
import { UnixFS } from 'ipfs-unixfs';
import toBuffer from 'it-to-buffer';
import { NotFoundError, NotUnixFSError } from '../../errors.js';
import { directoryContent } from "./content/directory.js";
import { fileContent } from "./content/file.js";
import { hamtShardedDirectoryContent } from "./content/hamt-sharded-directory.js";
const contentExporters = {
    raw: fileContent,
    file: fileContent,
    directory: directoryContent,
    'hamt-sharded-directory': hamtShardedDirectoryContent,
    metadata: (cid, node, unixfs, blockstore) => {
        return () => [];
    },
    symlink: (cid, node, unixfs, blockstore) => {
        return () => [];
    }
};
export async function dagPbResolver(cid, name, path, blockstore, options) {
    const block = await toBuffer(blockstore.get(cid, options));
    let node;
    try {
        node = decode(block);
    }
    catch (err) {
        // badly formatted or invalid protobuf
        throw new NotUnixFSError(err.message);
    }
    if (node.Data == null) {
        throw new NotUnixFSError('no data in PBNode');
    }
    let unixfs;
    try {
        unixfs = UnixFS.unmarshal(node.Data);
    }
    catch (err) {
        // non-UnixFS dag-pb node? It could happen.
        throw new NotUnixFSError(err.message);
    }
    const content = contentExporters[unixfs.type](cid, node, unixfs, path, blockstore);
    if (content == null) {
        throw new NotFoundError('could not find content exporter');
    }
    if (unixfs.isDirectory()) {
        return {
            type: 'directory',
            cid,
            name,
            path,
            entries: content,
            unixfs,
            node
        };
    }
    return {
        type: 'file',
        cid,
        name,
        path,
        content,
        unixfs,
        node,
        size: unixfs.fileSize()
    };
}
