// open-agents-nexus/dist/metering/index.js
/**
 * Usage metering engine
 *
 * Records per-request usage (tokens, bytes, duration, payment) and aggregates
 * summaries per peer per service. Supports hooks for audit/alerting.
 * FIFO eviction keeps memory bounded (default 10K records).
 */
import { createLogger } from '../logger.js';
const log = createLogger('metering');
const DEFAULT_MAX_RECORDS = 10_000;
// ---------------------------------------------------------------------------
// MeteringEngine
// ---------------------------------------------------------------------------
export class MeteringEngine {
    records = [];
    summaries = new Map();
    hooks = [];
    maxRecords;
    constructor(maxRecords = DEFAULT_MAX_RECORDS) {
        this.maxRecords = maxRecords;
    }
    /**
     * Record a usage event. Fires hooks, updates summaries, FIFO-evicts if needed.
     */
    async record(record) {
        this.records.push(record);
        // FIFO eviction
        while (this.records.length > this.maxRecords) {
            this.records.shift();
        }
        // Update peer summary
        this.updateSummary(record);
        // Fire hooks (errors are logged, not thrown)
        for (const hook of this.hooks) {
            try {
                await hook(record);
            }
            catch (err) {
                log.debug(`Metering hook error: ${err.message}`);
            }
        }
    }
    updateSummary(record) {
        let summary = this.summaries.get(record.peerId);
        if (!summary) {
            summary = {
                peerId: record.peerId,
                totalRequests: 0,
                totalInputBytes: 0,
                totalOutputBytes: 0,
                totalDurationMs: 0,
                totalTokens: 0,
                totalPayments: 0,
                services: new Map(),
                firstSeen: record.timestamp,
                lastSeen: record.timestamp,
            };
            this.summaries.set(record.peerId, summary);
        }
        summary.totalRequests++;
        summary.totalInputBytes += record.inputBytes;
        summary.totalOutputBytes += record.outputBytes;
        summary.totalDurationMs += record.durationMs;
        summary.totalTokens += record.tokens ?? 0;
        if (record.payment) {
            summary.totalPayments += parseFloat(record.payment.amount) || 0;
        }
        summary.lastSeen = Math.max(summary.lastSeen, record.timestamp);
        // Per-service breakdown
        const svcKey = record.service || record.capability;
        let svc = summary.services.get(svcKey);
        if (!svc) {
            svc = { requests: 0, inputBytes: 0, outputBytes: 0, durationMs: 0 };
            summary.services.set(svcKey, svc);
        }
        svc.requests++;
        svc.inputBytes += record.inputBytes;
        svc.outputBytes += record.outputBytes;
        svc.durationMs += record.durationMs;
    }
    /**
     * Add a hook that fires on every recorded usage event.
     */
    addHook(hook) {
        this.hooks.push(hook);
    }
    /**
     * Remove a previously added hook.
     */
    removeHook(hook) {
        const idx = this.hooks.indexOf(hook);
        if (idx >= 0)
            this.hooks.splice(idx, 1);
    }
    /**
     * Query records with optional filters.
     */
    getRecords(filter) {
        if (!filter)
            return [...this.records];
        return this.records.filter(r => {
            if (filter.peerId && r.peerId !== filter.peerId)
                return false;
            if (filter.service && r.service !== filter.service)
                return false;
            if (filter.capability && r.capability !== filter.capability)
                return false;
            if (filter.direction && r.direction !== filter.direction)
                return false;
            if (filter.since && r.timestamp < filter.since)
                return false;
            if (filter.until && r.timestamp > filter.until)
                return false;
            return true;
        });
    }
    /**
     * Get the aggregated usage summary for a peer.
     */
    getSummary(peerId) {
        return this.summaries.get(peerId);
    }
    /**
     * Get all peer summaries.
     */
    getAllSummaries() {
        return Array.from(this.summaries.values());
    }
    /**
     * Total number of records currently held.
     */
    get size() {
        return this.records.length;
    }
    /**
     * Clear all records and summaries.
     */
    clear() {
        this.records = [];
        this.summaries.clear();
    }
}
