// open-agents-nexus/dist/storage/checkpoint.js
import { createLogger } from '../logger.js';
const log = createLogger('storage:checkpoint');
// Create a checkpoint from a set of batch CIDs
export function createCheckpoint(roomId, epoch, aggregator, batchCids, previousCheckpoint = null) {
    return {
        schema: 'nexus:room-checkpoint:v1',
        roomId,
        epoch,
        createdAt: Date.now(),
        aggregator,
        batchCids,
        previousCheckpoint,
    };
}
// Sign a checkpoint
export async function signCheckpoint(checkpoint, privateKey) {
    const payload = JSON.stringify({
        schema: checkpoint.schema,
        roomId: checkpoint.roomId,
        epoch: checkpoint.epoch,
        createdAt: checkpoint.createdAt,
        aggregator: checkpoint.aggregator,
        batchCids: checkpoint.batchCids,
        previousCheckpoint: checkpoint.previousCheckpoint,
    });
    const sigBytes = await privateKey.sign(new TextEncoder().encode(payload));
    return { ...checkpoint, sig: Buffer.from(sigBytes).toString('base64') };
}
// Validate checkpoint structure
export function validateCheckpointStructure(cp) {
    if (!cp || typeof cp !== 'object')
        return false;
    const c = cp;
    if (c.schema !== 'nexus:room-checkpoint:v1')
        return false;
    if (typeof c.roomId !== 'string')
        return false;
    if (typeof c.epoch !== 'number')
        return false;
    if (typeof c.aggregator !== 'string')
        return false;
    if (!Array.isArray(c.batchCids))
        return false;
    if (typeof c.sig !== 'string')
        return false;
    return true;
}
// Suppress unused variable warning for logger in production code that may grow
void log;
