// @libp2p/webrtc/dist/src/private-to-public/transport.browser.js
import { serviceCapabilities, transportSymbol } from '@libp2p/interface';
import { peerIdFromString } from '@libp2p/peer-id';
import { CODE_P2P } from '@multiformats/multiaddr';
import { WebRTCDirect } from '@multiformats/multiaddr-matcher';
import { UnimplementedError } from "../error.js";
import { genUfrag } from '../util.js';
import { connect } from './utils/connect.js';
import { createDialerRTCPeerConnection } from './utils/get-rtcpeerconnection.js';
export class WebRTCDirectTransport {
    log;
    metrics;
    components;
    init;
    constructor(components, init = {}) {
        this.log = components.logger.forComponent('libp2p:webrtc-direct');
        this.components = components;
        this.init = init;
        if (components.metrics != null) {
            this.metrics = {
                dialerEvents: components.metrics.registerCounterGroup('libp2p_webrtc-direct_dialer_events_total', {
                    label: 'event',
                    help: 'Total count of WebRTC-direct dial events by type'
                })
            };
        }
    }
    [transportSymbol] = true;
    [Symbol.toStringTag] = '@libp2p/webrtc-direct';
    [serviceCapabilities] = [
        '@libp2p/transport'
    ];
    /**
     * Dial a given multiaddr
     */
    async dial(ma, options) {
        this.log('dial %a', ma);
        // do not create RTCPeerConnection if the signal has already been aborted
        options.signal.throwIfAborted();
        let theirPeerId;
        const remotePeerString = ma.getComponents().findLast(c => c.code === CODE_P2P)?.value;
        if (remotePeerString != null) {
            theirPeerId = peerIdFromString(remotePeerString);
        }
        const ufrag = genUfrag();
        // https://github.com/libp2p/specs/blob/master/webrtc/webrtc-direct.md#browser-to-public-server
        const { peerConnection, muxerFactory } = await createDialerRTCPeerConnection('client', ufrag, {
            rtcConfiguration: typeof this.init.rtcConfiguration === 'function' ? await this.init.rtcConfiguration() : this.init.rtcConfiguration ?? {},
            dataChannel: this.init.dataChannel
        });
        try {
            return await connect(peerConnection, muxerFactory, ufrag, {
                role: 'client',
                log: this.log,
                logger: this.components.logger,
                events: this.metrics?.dialerEvents,
                signal: options.signal,
                remoteAddr: ma,
                dataChannel: this.init.dataChannel,
                upgrader: options.upgrader,
                peerId: this.components.peerId,
                remotePeer: theirPeerId,
                privateKey: this.components.privateKey
            });
        }
        catch (err) {
            peerConnection.close();
            throw err;
        }
    }
    /**
     * Create a transport listener - this will throw in browsers
     */
    createListener(options) {
        throw new UnimplementedError('WebRTCDirectTransport.createListener');
    }
    /**
     * Filter check for all Multiaddrs that this transport can listen on
     */
    listenFilter(multiaddrs) {
        return [];
    }
    /**
     * Filter check for all Multiaddrs that this transport can dial
     */
    dialFilter(multiaddrs) {
        return multiaddrs.filter(WebRTCDirect.exactMatch);
    }
}
