// @libp2p/http/dist/src/middleware/cookies.js
import { toURL } from '@libp2p/http-utils';
import cookie from 'cookie';
export class Cookies {
    log;
    cookies;
    constructor(components, init = {}) {
        this.log = components.logger.forComponent('libp2p:http:cookies');
        this.cookies = new Map();
    }
    async prepareRequest(resource, opts) {
        const credentials = opts.credentials ?? 'same-origin';
        if (credentials === 'omit') {
            return;
        }
        const origin = opts.headers.get('origin');
        if (origin == null || origin === 'null') {
            return;
        }
        const url = toURL(resource, opts.headers);
        const cookies = (this.cookies.get(url.hostname) ?? []).filter(cookie => {
            if (cookie.expires != null && cookie.expires < Date.now()) {
                return false;
            }
            if (cookie.path != null && !url.pathname.startsWith(cookie.path)) {
                return false;
            }
            return true;
        })
            .map(cookie => `${cookie.name}=${cookie.value}`)
            .join('; ');
        if (cookies.length > 0) {
            opts.headers.set('cookie', cookies);
        }
    }
    async processResponse(resource, opts, response) {
        const credentials = opts.credentials ?? 'same-origin';
        if (credentials === 'omit') {
            removeSetCookie(response);
            return;
        }
        const origin = opts.headers.get('origin');
        if (origin == null || origin === 'null') {
            return;
        }
        const url = toURL(resource, opts.headers);
        for (const value of response.headers.getSetCookie()) {
            const cookies = [
                ...(this.cookies.get(url.hostname) ?? []),
                ...toCookies(cookie.parse(value))
            ];
            this.cookies.set(url.hostname, cookies);
        }
        removeSetCookie(response);
    }
}
/**
 * the fetch spec requires not exposing set-cookie to client code
 */
function removeSetCookie(response) {
    if (response.headers.has('set-cookie')) {
        response.headers.delete('set-cookie');
    }
    return response;
}
function toCookies(parsed) {
    const metadata = {};
    const output = [];
    Object.entries(parsed).forEach(([name, value]) => {
        if (name.toLowerCase() === 'domain' && value != null) {
            metadata.domain = value;
        }
        if (name.toLowerCase() === 'max-age' && value != null) {
            metadata.expires = Date.now() + (parseInt(value, 10) * 1000);
        }
        if (!COOKIE_FIELDS.includes(name.toLowerCase()) && value != null) {
            output.push({
                name,
                value
            });
        }
    });
    return output.map(c => ({
        ...c,
        ...metadata
    }));
}
const COOKIE_FIELDS = [
    'domain',
    'expires',
    'httponly',
    'max-age',
    'partitioned',
    'path',
    'samesite',
    'secure'
];
