// @libp2p/http/dist/src/middleware/peer-id-auth.js
import { MissingAuthHeaderError, DEFAULT_AUTH_TOKEN_TTL, ClientInitiatedHandshake } from '@libp2p/http-peer-id-auth';
import { getHost, isWebSocketUpgrade } from '@libp2p/http-utils';
import { InvalidMessageError, InvalidParametersError } from '@libp2p/interface';
import { CODE_P2P } from '@multiformats/multiaddr';
export const SEC_WEBSOCKET_PROTOCOL_PREFIX = 'authorization=';
export class PeerIdAuth {
    components;
    tokens;
    tokenTTL;
    verifyPeer;
    constructor(components, init) {
        this.components = components;
        this.tokens = new Map();
        this.tokenTTL = init.tokenTTL ?? DEFAULT_AUTH_TOKEN_TTL;
        this.verifyPeer = init.verifyPeer;
    }
    async prepareRequest(resource, opts) {
        const authorization = await this.getOrCreateAuthToken(resource, opts);
        if (isWebSocketInit(opts)) {
            opts.protocols = [
                ...(opts.protocols ?? []),
                `${SEC_WEBSOCKET_PROTOCOL_PREFIX}${btoa(authorization)}`
            ];
        }
        else {
            if (opts.headers.get('authorization') != null) {
                throw new InvalidParametersError('Will not overwrite existing Authorization header');
            }
            opts.headers.set('Authorization', authorization);
        }
    }
    async getOrCreateAuthToken(resource, opts) {
        const key = getCacheKey(resource, opts.headers);
        let token = this.tokens.get(key);
        // check token expiry
        if (token?.expires != null && token?.expires < Date.now()) {
            this.tokens.delete(key);
            token = undefined;
        }
        // create new token
        if (token == null) {
            return this.createAuthToken(resource, opts);
        }
        return token?.authorization;
    }
    async createAuthToken(resource, opts) {
        const hostname = getHost(resource, opts.headers);
        const handshake = new ClientInitiatedHandshake(this.components.privateKey, hostname, this.verifyPeer);
        const challenge = handshake.getChallenge();
        // copy existing headers
        const challengeHeaders = new Headers(opts.headers);
        challengeHeaders.set('authorization', challenge);
        // remove any WebSocket upgrade headers
        challengeHeaders.delete('connection');
        challengeHeaders.delete('upgrade');
        // get the server's response
        const resp = await this.components.http.fetch(resource, {
            method: 'OPTIONS',
            headers: challengeHeaders,
            signal: opts.signal,
            middleware: opts.middleware
                .filter(m => m !== this)
                .map(m => () => m)
        });
        // verify the server's challenge
        const response = resp.headers.get('www-authenticate');
        if (response == null) {
            throw new MissingAuthHeaderError('No www-authenticate header in response');
        }
        // verify remote server and answer the server challenge
        const authorization = await handshake.verifyServer(response, opts);
        if (handshake.serverId == null) {
            throw new InvalidMessageError('Failed to get server PeerId');
        }
        const key = getCacheKey(resource, opts.headers);
        this.tokens.set(key, {
            peerId: handshake.serverId,
            authorization,
            expires: Date.now() + this.tokenTTL,
            handshake
        });
        return authorization;
    }
    processResponse(resource, opts, response) {
        const key = getCacheKey(resource, opts.headers);
        const token = this.tokens.get(key);
        // add the remote peer id as a response header
        if (token?.peerId != null) {
            const headers = new Headers(response.headers);
            headers.set('x-libp2p-peer-id', token.peerId.toString());
            // the headers property is read-only so we can't just re-assign it
            response = new Response(response.body, {
                status: response.status,
                headers
            });
        }
        // store the bearer token if the server provided it
        const serverAuthHeader = response.headers.get('authentication-info');
        if (serverAuthHeader != null && token != null) {
            token.authorization = token.handshake.decodeBearerToken(serverAuthHeader);
        }
    }
}
export function peerIdAuth(init = {}) {
    return (components) => {
        return new PeerIdAuth(components, init);
    };
}
function isWebSocketInit(opts) {
    if (opts == null) {
        return false;
    }
    return isWebSocketUpgrade(opts.method, opts.headers);
}
function getCacheKey(resource, headers) {
    let prefix = '';
    if (Array.isArray(resource)) {
        const peer = resource.map(ma => ma.getComponents().findLast(c => c.code === CODE_P2P)?.value)
            .filter(Boolean)
            .pop();
        if (peer != null) {
            prefix = `${peer}-`;
        }
    }
    return `${prefix}${getHost(resource, headers)}`;
}
