// open-agents-nexus/dist/nkn/index.js
import { createLogger } from '../logger.js';
const log = createLogger('nkn');
export const DEFAULT_NKN_CONFIG = {
    enabled: false, // disabled by default — opt-in fallback
    identifier: 'nexus',
    numSubClients: 3,
};
export class NknFallback {
    client = null;
    config;
    addressInfo = null;
    onMessage = null;
    constructor(config) {
        this.config = { ...DEFAULT_NKN_CONFIG, ...config };
    }
    // Generate or load an NKN address
    async connect(seed) {
        if (!this.config.enabled)
            return null;
        try {
            const nkn = await import('nkn-sdk');
            const clientOpts = {
                identifier: this.config.identifier,
                numSubClients: this.config.numSubClients,
            };
            if (seed)
                clientOpts.seed = seed;
            this.client = new nkn.MultiClient(clientOpts);
            // Wait for ready
            await new Promise((resolve, reject) => {
                const timeout = setTimeout(() => reject(new Error('NKN connect timeout')), 15000);
                this.client.onConnect(() => {
                    clearTimeout(timeout);
                    resolve();
                });
            });
            this.addressInfo = {
                address: this.client.addr,
                publicKey: this.client.getPublicKey(),
                seed: this.client.getSeed(),
            };
            log.info(`NKN connected: ${this.addressInfo.address}`);
            // Listen for messages
            this.client.onMessage(({ src, payload }) => {
                this.onMessage?.(src, typeof payload === 'string' ? payload : new TextDecoder().decode(payload));
            });
            return this.addressInfo;
        }
        catch (err) {
            log.warn(`NKN connection failed (fallback disabled): ${err}`);
            return null;
        }
    }
    // Send a message to an NKN address
    async send(to, message) {
        if (!this.client)
            return false;
        try {
            await this.client.send(to, message);
            return true;
        }
        catch (err) {
            log.debug(`NKN send failed: ${err}`);
            return false;
        }
    }
    // Set message handler
    setMessageHandler(handler) {
        this.onMessage = handler;
    }
    async disconnect() {
        if (this.client) {
            try {
                await this.client.close();
            }
            catch { /* ignore */ }
            this.client = null;
            this.addressInfo = null;
            log.info('NKN disconnected');
        }
    }
    get address() {
        return this.addressInfo?.address ?? null;
    }
    get isConnected() {
        return this.client !== null;
    }
}
