// open-agents-nexus/dist/sponsor/index.js
/**
 * sponsor/index.ts — COHERE Plane 3: Sponsor Pools and Free-Tier Funding
 *
 * Implements treasury management for subsidized inference per the unified
 * COHERE paper (p.43-44). The free tier is a network growth engine, not charity.
 *
 * Revenue split (COHERE proposal, p.46):
 *   70% provider payout
 *   10% commons subsidy pool
 *    8% memory and storage stewardship
 *    7% relay/discovery/bandwidth
 *    5% protocol reserve and dispute handling
 *
 * Provenance: Project COHERE Unified Complete §Free sponsored inference
 */
import { createLogger } from '../logger.js';
const log = createLogger('sponsor');
// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------
/** Revenue split percentages per COHERE paper p.46 */
export const COHERE_REVENUE_SPLIT = {
    provider: 0.70,
    commons: 0.10,
    memory: 0.08,
    relay: 0.07,
    reserve: 0.05,
};
// ---------------------------------------------------------------------------
// SponsorPoolManager
// ---------------------------------------------------------------------------
export class SponsorPoolManager {
    pool;
    freeUsers = new Map();
    contributions = [];
    constructor(initialBalance = 0, dailyBudget = 100) {
        const now = new Date();
        this.pool = {
            balance: initialBalance,
            dailyBudget,
            dailySpent: 0,
            epoch: 1,
            epochResetAt: now.toISOString(),
            classBudgets: {
                'commons-chat': dailyBudget * 0.4,
                'public-knowledge': dailyBudget * 0.3,
                'network-maintenance': dailyBudget * 0.15,
                'community-research': dailyBudget * 0.15,
            },
            classSpent: {
                'commons-chat': 0,
                'public-knowledge': 0,
                'network-maintenance': 0,
                'community-research': 0,
            },
        };
    }
    /** Check if a free request can be funded */
    canFundRequest(requestClass, estimatedCost) {
        if (this.pool.dailySpent + estimatedCost > this.pool.dailyBudget)
            return false;
        if (this.pool.classSpent[requestClass] + estimatedCost > this.pool.classBudgets[requestClass])
            return false;
        if (this.pool.balance < estimatedCost)
            return false;
        return true;
    }
    /** Reserve funds for a sponsored request */
    reserveFunds(requestClass, amount) {
        if (!this.canFundRequest(requestClass, amount))
            return false;
        this.pool.balance -= amount;
        this.pool.dailySpent += amount;
        this.pool.classSpent[requestClass] += amount;
        log.info(`Sponsor pool: reserved $${amount.toFixed(4)} for ${requestClass} (remaining: $${this.pool.balance.toFixed(4)})`);
        return true;
    }
    /** Deposit funds into the sponsor pool */
    deposit(amount, source) {
        this.pool.balance += amount;
        log.info(`Sponsor pool: +$${amount.toFixed(4)} from ${source} (total: $${this.pool.balance.toFixed(4)})`);
    }
    /** Apply revenue split from a paid request */
    applyRevenueSplit(totalRevenue) {
        const split = {
            provider: totalRevenue * COHERE_REVENUE_SPLIT.provider,
            commons: totalRevenue * COHERE_REVENUE_SPLIT.commons,
            memory: totalRevenue * COHERE_REVENUE_SPLIT.memory,
            relay: totalRevenue * COHERE_REVENUE_SPLIT.relay,
            reserve: totalRevenue * COHERE_REVENUE_SPLIT.reserve,
        };
        // Auto-deposit commons portion into sponsor pool
        this.pool.balance += split.commons;
        return split;
    }
    /** Get or create a free-user envelope */
    getFreeUserEnvelope(peerId) {
        let env = this.freeUsers.get(peerId);
        if (!env) {
            env = {
                baseTokens: 10000, // default daily token budget for free users
                tokensUsed: 0,
                queuePriority: 'low',
                maxModelTier: 'medium',
                contributorBonus: 0,
            };
            this.freeUsers.set(peerId, env);
        }
        return env;
    }
    /** Record a contribution credit */
    recordContribution(credit) {
        this.contributions.push(credit);
        // Boost the contributor's free-user envelope
        const env = this.getFreeUserEnvelope(credit.peerId);
        env.contributorBonus += credit.creditEarned;
        env.queuePriority = 'normal'; // contributors get normal priority
        log.info(`Contribution credit: ${credit.peerId} earned ${credit.creditEarned} tokens for ${credit.type}`);
    }
    /** Reset epoch (call at daily boundary) */
    resetEpoch() {
        this.pool.epoch++;
        this.pool.dailySpent = 0;
        this.pool.epochResetAt = new Date().toISOString();
        for (const key of Object.keys(this.pool.classSpent)) {
            this.pool.classSpent[key] = 0;
        }
        // Reset free-user daily usage
        for (const env of this.freeUsers.values()) {
            env.tokensUsed = 0;
        }
        log.info(`Sponsor pool: epoch ${this.pool.epoch} started`);
    }
    /** Get pool state for reporting */
    getState() {
        return { ...this.pool };
    }
    /** Get all contribution credits */
    getContributions() {
        return this.contributions;
    }
}
