// open-agents-nexus/dist/trust/default-policy.js
import { createLogger } from '../logger.js';
const log = createLogger('trust');
export class DefaultTrustPolicy {
    allowSet;
    denySet;
    roomDenySet;
    constructor(config = {}) {
        this.allowSet = new Set(config.allowlist ?? []);
        this.denySet = new Set(config.denylist ?? []);
        this.roomDenySet = new Set(config.roomDenylist ?? []);
    }
    allowPeer(peerId) {
        if (this.denySet.has(peerId)) {
            log.warn(`Peer denied: ${peerId.slice(0, 12)}...`);
            return false;
        }
        // If allowlist is non-empty, only allow listed peers
        if (this.allowSet.size > 0 && !this.allowSet.has(peerId)) {
            return false;
        }
        return true;
    }
    allowRoom(roomId, peerId) {
        if (!this.allowPeer(peerId))
            return false;
        if (this.roomDenySet.has(roomId))
            return false;
        return true;
    }
    allowCapability(peerId, capability) {
        return this.allowPeer(peerId);
    }
    allowRelay(peerId) {
        return this.allowPeer(peerId);
    }
    // Dynamic mutation
    addToAllowlist(peerId) { this.allowSet.add(peerId); }
    removeFromAllowlist(peerId) { this.allowSet.delete(peerId); }
    addToDenylist(peerId) { this.denySet.add(peerId); }
    removeFromDenylist(peerId) { this.denySet.delete(peerId); }
    denyRoom(roomId) { this.roomDenySet.add(roomId); }
    allowRoomId(roomId) { this.roomDenySet.delete(roomId); }
}
